### PR TITLE
test(python): Test S3 functionality using moto server

### PR DESCRIPTION
**Pull request by cjackal**
As requested by @ritchie46 in #10008

### CODE REVIEW

File: py-polars/tests/unit/io/test_cloud.py [hunk 1]
```diff
@@ -11,3 +52,53 @@ def test_err_on_s3_glob() -> None:
...
+
+
+@pytest.mark.usefixtures("_s3")
+def test_scan_parquet_on_s3() -> None:
+    df = pl.scan_parquet(
+        "s3://bucket/foods1.parquet",
+        storage_options={"endpoint_url": f"http://{host}:{port}"},
+    )
+    assert df.columns == ["category", "calories", "fats_g", "sugars_g"]
+    assert df.collect().shape == (27, 4)
```
Comment (cjackal) [inline py-polars/tests/unit/io/test_cloud.py:104]: Seems `scan_*` are yet to be working properly; tests succeeded on schema reading part, but failed on `.collect()` part. This is not `moto`'s fault, `scan_parquet` currently fails on master too.

(try `pl.scan_parquet("s3://saturn-public-data/nyc-taxi/data/yellow_tripdata_2019-01.parquet").collect()` and you will see)

File: py-polars/requirements-dev.txt [hunk 2]
```diff
@@ -19,10 +19,14 @@ adbc_driver_sqlite; python_version >= '3.9' and platform_system != 'Windows'
...
 cloudpickle
 fsspec
+s3fs[boto3]
 
 # Tooling
+flask!=2.2.0,!=2.2.1  # Required for moto.server w/o installing all moto[server] dependencies
+flask-cors  # Required for moto.server w/o installing all moto[server] dependencies
 hypothesis==6.82.0
 maturin==1.1.0
+moto[s3]==4.1.13  # Need moto.server to mock s3fs - see aio-libs/aiobotocore#755
```
Comment (cjackal) [inline py-polars/requirements-dev.txt:29]: As commented here, we need `moto.server`, not only `moto[s3]`. But `moto[server]` is a total dependency mess due to cloudwatch support - especially it pins `pydantic` to `~1.8`, breaking other tests. We only need S3 part of the server, a brainless pass of manually install `flask` is taken for now.

**Comment by cjackal** [comment]
Our new unit tests in this PR already played its role in detecting #10174 🎊

**Comment by cjackal** [comment]
New test suits now pass after #10175

**Comment by cjackal** [comment]
A humble ping to reviewers @ritchie46 @stinodego @alexander-beedie that this PR is ready for review.

I was a little worried if this test is slow (perhaps too slow to linger overall CI workflow?), but it turns out to be negligible (a few seconds of launching a server is marginal within running all ~3000 tests)

File: py-polars/tests/unit/io/test_cloud.py [hunk 3]
```diff
@@ -11,3 +52,53 @@ def test_err_on_s3_glob() -> None:
...
+def test_read_parquet_on_s3() -> None:
+    df = pl.read_parquet(
+        "s3://bucket/foods1.parquet",
+        storage_options={"endpoint_url": f"http://{host}:{port}"},
+    )
+    assert df.columns == ["category", "calories", "fats_g", "sugars_g"]
+    assert df.shape == (27, 4)
+
+
+@pytest.mark.usefixtures("_s3")
```
Comment (ritchie46) [inline py-polars/tests/unit/io/test_cloud.py:97]: can you mark these as `slow`. 

That will make sure they only run in CI and we don't have to launch the server locally unless we want to.

File: py-polars/requirements-dev.txt (hunk 2 above)
Comment (ritchie46) [reply py-polars/requirements-dev.txt:29]: Could we maybe elide this problem with a moto github action? https://github.com/getmoto/moto

**Review by ritchie46 — COMMENTED** [review]
Thanks for the PR! I have left some comments.

File: py-polars/tests/unit/io/test_cloud.py (hunk 3 above)
Comment (cjackal) [reply py-polars/tests/unit/io/test_cloud.py:102]: Thought the same, all moto tests now have been `@pytest.mark.slow`ed.

File: py-polars/requirements-dev.txt (hunk 2 above)
Comment (cjackal) [reply py-polars/requirements-dev.txt:29]: I've thought a bit and feels like I don't follow what you mean, do you mean launching moto server outside test venv (w/ moto github action) and removing moto dependencies from `requirements-dev.txt`? Doesn't it make `test_cloud.py` not running under local dev setup?

**Comment by ritchie46** [comment]
Thanks a lot for this work @cjackal. I think this is very valuable. I want to leave the final review to @stinodego as he is our CI expert.

**Comment by stinodego** [comment]
I'll test this out later today or tomorrow!

**Comment by cjackal** [comment]
@stinodego Just a comment on latest nontrivial commit:
As we are running `pytest --dist loadgroup`, I grouped `moto`-related test suits as `cloud` tag so that they are running in a single worker. It would both speed up the test running time and avoid launching moto server multiple times (in case of multiple workers).

**Review by stinodego — APPROVED** [review]
I did some minor refactoring - this should now be good to go!

Thanks for the effort @cjackal !

**Pull request merged**